lib/ecto/multi.ex

```diff
@@ -0,0 +1,195 @@
+      raise """
+      when merging following Ecto.Multi:
+
+      #{inspect lhs}
+
+      #{inspect rhs}
+
+      both declared operations: #{inspect common}
+      """
+    end
+  end
+
+  @spec insert(t, name, Changeset.t | Ecto.Schema.t) :: t
+  def insert(multi, name, changeset_or_struct, opts \\ [])
+
+  def insert(multi, name, %Changeset{} = changeset, opts) do
+    add_changeset(multi, :insert, name, changeset, opts)
+  end
+
+  def insert(multi, name, struct, opts) do
+    insert(multi, name, Changeset.change(struct), opts)
+  end
+
+  @spec update(t, name, Changeset.t) :: t
+  def update(multi, name, %Changeset{} = changeset, opts \\ []) do
+    add_changeset(multi, :update, name, changeset, opts)
+  end
+
+  @spec delete(t, name, Changeset.t | Ecto.Schema.t) :: t
+  def delete(multi, name, changeset_or_struct, opts \\ [])
+
+  def delete(multi, name, %Changeset{} = changeset, opts) do
+    add_changeset(multi, :delete, name, changeset, opts)
+  end
+
+  def delete(multi, name, struct, opts) do
+    delete(multi, name, Changeset.change(struct), opts)
+  end
+
+  defp add_changeset(multi, action, name, changeset, opts) when is_list(opts) do
+    add_operation(multi, name, {:changeset, put_action(changeset, action), opts})
+  end
+
+  defp put_action(%{action: nil} = changeset, action) do
+    %{changeset | action: action}
+  end
+
+  defp put_action(%{action: action} = changeset, action) do
+    changeset
+  end
+
+  defp put_action(%{action: original}, action) do
+    raise ArgumentError, "you provided a changeset with an action already set " <>
+      "to #{inspect original} when trying to #{action} it"
+  end
+
+  @spec run(t, name, run) :: t
+  def run(multi, name, run) when is_function(run, 1) do
+    add_operation(multi, name, {:run, run})
+  end
+
+  @spec run(t, name, module, function, args) :: t
+    when function: atom, args: [any]
+  def run(multi, name, mod, fun, args) do
+    add_operation(multi, name, {:run, {mod, fun, args}})
+  end
+
+  @spec insert_all(t, name, schema_or_source, [entry], Keyword.t) :: t
+    when schema_or_source: binary | {binary | nil, binary} | Ecto.Schema.t,
+         entry: map | Keyword.t
+  def insert_all(multi, name, schema_or_source, entries, opts \\ []) when is_list(opts) do
+    add_operation(multi, name, {:insert_all, schema_or_source, entries, opts})
+  end
+
+  @spec update_all(t, name, Ecto.Queryable.t, Keyword.t, Keyword.t) :: t
+  def update_all(multi, name, queryable, updates, opts \\ []) when is_list(opts) do
+    query = Ecto.Queryable.to_query(queryable)
+    add_operation(multi, name, {:update_all, query, updates, opts})
+  end
+
+  @spec delete_all(t, name, Ecto.Queryable.t, Keyword.t) :: t
+  def delete_all(multi, name, queryable, opts \\ []) when is_list(opts) do
+    query = Ecto.Queryable.to_query(queryable)
+    add_operation(multi, name, {:delete_all, query, opts})
+  end
+
+  defp add_operation(%Multi{operations: operations, names: names} = multi, name,
```

Comment:
Same can be done there: https://github.com/elixir-lang/ecto/pull/1138/files#diff-290dd885b452ddaec406658d1c820044R157.